File: app/src/main/java/com/daniel/mobilepauker2/dropbox/DropboxClientFactory.java
package com.daniel.mobilepauker2.dropbox;

import com.dropbox.core.DbxRequestConfig;
import com.dropbox.core.http.OkHttp3Requestor;
import com.dropbox.core.v2.DbxClientV2;

/**
 * Singleton instance of {@link DbxClientV2} and friends
 */
class DropboxClientFactory {

    private static DbxClientV2 sDbxClient;

    static void init(String accessToken) {
        if (sDbxClient == null) {
            DbxRequestConfig requestConfig = DbxRequestConfig.newBuilder("examples-v2-demo")
                .withHttpRequestor(new OkHttp3Requestor(OkHttp3Requestor.defaultOkHttpClient()))
                .build();

            sDbxClient = new DbxClientV2(requestConfig, accessToken);
        }
    }

    static DbxClientV2 getClient() {
        if (sDbxClient == null) {
            throw new IllegalStateException("Client not initialized.");
        }
        return sDbxClient;
    }
}

File: app/src/main/java/de/daniel/mobilepauker2/dropbox/DropboxClientFactory.kt
package de.daniel.mobilepauker2.dropbox

import com.dropbox.core.DbxRequestConfig
import com.dropbox.core.http.OkHttp3Requestor
import com.dropbox.core.oauth.DbxCredential
import com.dropbox.core.v2.DbxClientV2
import de.daniel.mobilepauker2.utils.Constants

/**
 * Singleton instance of [DbxClientV2] and friends
 */
internal object DropboxClientFactory {
    private var sDbxClient: DbxClientV2? = null

    fun init(credential: DbxCredential) {
        val newCredential = DbxCredential(
            credential.accessToken,
            -1L,
            credential.refreshToken,
            credential.appKey
        )
        if (sDbxClient == null) {
            sDbxClient = DbxClientV2(DbxRequestConfigFactory.requestConfig, newCredential)
        }
    }

    fun readCredentialFromString(credential: String): DbxCredential =
        DbxCredential.Reader.readFully(credential)

    val client: DbxClientV2
        get() {
            checkNotNull(sDbxClient) { "Client not initialized." }
            return sDbxClient!!
        }
}